### PR TITLE
[Serializer] Fix denormalization of object with variadic constructor typed argument 

### PR DESCRIPTION
**Pull request by ajgarlag**
| Q             | A
| ------------- | ---
| Branch?       | 3.4 up to 4.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31436
| License       | MIT

This PR adds a test to demonstrate the bug, and a fix to squash it.

### CODE REVIEW

**Comment by ajgarlag** [comment]
PR ready to merge

File: src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php [hunk 1]
```diff
@@ -128,4 +131,19 @@ public function testObjectWithNullableConstructorArgument()
 
         $this->assertNull($dummy->getFoo());
     }
+
+    /**
+     * @requires PHP 7.1
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php]: Shouldn't it be 5.6?

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 2]
```diff
@@ -354,7 +354,15 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
                             throw new RuntimeException(sprintf('Cannot create an instance of %s from serialized data because the variadic parameter %s can only accept an array.', $class, $constructorParameter->name));
                         }
 
-                        $params = array_merge($params, $data[$paramName]);
+                        $variadicParameters = array_map(
+                            function ($parameterData) use ($reflectionClass, $constructorParameter, $paramName, $context, $format) {
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Can you use a foreach loop instead of `array_map`? It's a bit faster, and it can matter here because it's on the hot path.

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php [hunk 3]
```diff
@@ -354,7 +354,15 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
...
+                                return $this->denormalizeParameter($reflectionClass, $constructorParameter, $paramName, $parameterData, $context, $format);
+                            },
+                            $data[$paramName]
+                        );
+
+                        $params = array_merge($params, $variadicParameters);
+                        unset($data[$key]);
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Why this change?

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php (hunk 1 above)
Comment (ajgarlag) [reply src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php]: You are right!

File: src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php (hunk 3 above)
Comment (ajgarlag) [reply src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php]: Without this changes, when using the `PropertyNormalizer`, the deserialized object is successfully instantiated, but the property is overwritten again with the original `$data[$key]` array value ignoring the type hint.

This is tested in https://github.com/symfony/symfony/pull/31438/files#diff-9d3a423164acd48c3fd313cff0be945bR146

File: src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php (hunk 1 above)
Comment (ajgarlag) [reply src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php]: After testing, it fails in 5.6 because I'm using `NullableConstructorArgumentDummy` (which requires PHP 7.1) as type hint in the fixture. It is not detected by travis-ci.org because PHP 5.6 is not being used to test PRs.

Comment (ajgarlag) [reply src/Symfony/Component/Serializer/Tests/Normalizer/AbstractNormalizerTest.php]: Fixed using `Dummy` insteadof `NullableConstructorArgumentDummy`. Tests finally passes in PHP 5.6.

**Review by dunglas — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @ajgarlag.

**Comment by ajgarlag** [comment]
@nicolas-grekas I'd like to have this issue fixed in 4.2 branch too. Can you tell me when will be merged into that branch? Thanks.

**Comment by nicolas-grekas** [comment]
There is nothing to do on your side, we will merge 3.4 into 4.2 with the fix.

**Comment by WalentinG** [comment]
@nicolas-grekas, the fix breaks current workaround 
try dummy object below in your test
```php
<?php

namespace Symfony\Component\Serializer\Tests\Fixtures;

final class VariadicConstructorAnnotationTypedArgsDummy
{
    /** @var Dummy[] */
    private $foo;
    
    public function __construct(...$foo)
    {
        $this->foo = $foo;
    }
    
    public function getFoo()
    {
        return $this->foo;
    }
}
```
